### PR TITLE
CORTX-28056: Update IAM User

### PR DESCRIPTION
**Pull request by shalakadharap**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -2990,6 +3023,20 @@ int MotrStore::store_access_key(const DoutPrefixProvider *dpp, optional_yield y,
   return rc;
 }
 
+int MotrStore::store_email_key(const DoutPrefixProvider *dpp, optional_yield y, MotrEmailInfo emailInfo)
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:3026]: `store_email_info()` looks better choice for function name here.

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -2990,6 +3023,20 @@ int MotrStore::store_access_key(const DoutPrefixProvider *dpp, optional_yield y,
...
+{
+  int rc;
+  bufferlist bl;
+  emailInfo.encode(bl);
+  rc = do_idx_op_by_name(RGW_IAM_MOTR_EMAIL_KEY,
+                                M0_IC_PUT, emailInfo.email_id, bl);
+  if (rc < 0){
+    ldout(cctx, 0) << "Failed to store key: rc = " << rc << dendl;
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:3034]: Please update key -> email

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -400,6 +401,9 @@ int MotrUser::store_user(const DoutPrefixProvider* dpp,
     store->store_access_key(dpp, y, MGWUserKeys);
   }
 
+  MotrEmailInfo emailInfo(info.user_email, info.user_id.id);
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:404]: Please put an empty check here. We should store anything in email index table only if email is supplied while user creation.

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.cc:3034]: Didnt understand. Have passed email_id as key only

Comment (jjxsg) [reply src/rgw/rgw_sal_motr.cc:3034]: Key here represents access key. Error should indicate what it failed to store.

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -2967,7 +2970,34 @@ int MotrStore::get_user_by_access_key(const DoutPrefixProvider *dpp, const std::
...
+  rc = do_idx_op_by_name(RGW_IAM_MOTR_EMAIL_KEY,
+                           M0_IC_GET, email, bl);
+  if (rc < 0){
+    ldout(cctx, 0) << "Email Id not found: rc = " << rc << dendl;
+    return rc;
+  }
+
+  bufferlist& blr = bl;
+  auto iter = blr.cbegin();
+  uinfo.decodeUserByEmail(iter);
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:2987]: For function names, Camel Casing is not used in ceph I think.
`decode_user_by_email` might be better.

File: src/rgw/rgw_sal_motr.h [hunk 5]
```diff
@@ -995,6 +1013,7 @@ class MotrStore : public Store {
                           std::string key_str, bufferlist &bl, bool update=true);
     int check_n_create_global_indices();
     int store_access_key(const DoutPrefixProvider *dpp, optional_yield y, MotrAccessKey access_key);
+    int store_email_key(const DoutPrefixProvider *dpp, optional_yield y, std::string email);
```
[the remark this answers is not in the document]
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.h:1016]: Fixed

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.cc:3031]: Not changing this

File: src/rgw/rgw_sal_motr.cc (hunk 4 above)
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.cc:2987]: Fixed

File: src/rgw/rgw_sal_motr.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.h]: Fixed.
New name- store_email_info()

**Review by sumedhak27 — APPROVED** [review]

**Comment by jjxsg** [comment]
I only see changes for `get_user_by_email()` 
Are all other update commands for IAM user working after enabling the above function?

**Review by jjxsg — CHANGES_REQUESTED** [review]
Also please follow the guidelines for PRs and commits suggested by @andriytk

File: src/rgw/rgw_sal_motr.cc [hunk 6]
```diff
@@ -399,7 +400,9 @@ int MotrUser::store_user(const DoutPrefixProvider* dpp,
     MotrAccessKey MGWUserKeys(access_key, secret_key, info.user_id.id);
     store->store_access_key(dpp, y, MGWUserKeys);
   }
-
+ if(!info.user_email.empty()) {
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:403]: Style: add space after `if` and before `(`.

**Review by andriytk — CHANGES_REQUESTED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 6 above)
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.cc:403]: Fixed

File: src/rgw/rgw_sal_motr.cc [hunk 7]
```diff
@@ -2990,6 +3020,20 @@ int MotrStore::store_access_key(const DoutPrefixProvider *dpp, optional_yield y,
...
 }
 
+int MotrStore::store_email_info(const DoutPrefixProvider *dpp, optional_yield y, std::string email_id)
+{
+  int rc;
+  bufferlist bl;
+  info.encode_user_by_email(bl);
+  rc = do_idx_op_by_name(RGW_IAM_MOTR_EMAIL_KEY,
+                                M0_IC_PUT, email_id, bl);
+  if (rc < 0){
```
[the remark this answers is not in the document]
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.cc:3030]: Fixed

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.cc:3031]: Fixed

File: src/rgw/rgw_sal_motr.cc [hunk 8]
```diff
@@ -2990,6 +3020,20 @@ int MotrStore::store_access_key(const DoutPrefixProvider *dpp, optional_yield y,
...
+    ldout(cctx, 0) << "Failed to store key: rc = " << rc << dendl;
+    return rc;
```
[the remark this answers is not in the document]
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.cc:3032]: why?

File: src/rgw/rgw_sal_motr.h [hunk 9]
```diff
@@ -163,6 +180,7 @@ struct MotrAccessKey {
 };
 WRITE_CLASS_ENCODER(MotrAccessKey);
 
+
```
[the remark this answers is not in the document]
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.h:183]: Fixed

File: src/rgw/rgw_sal_motr.cc [hunk 10]
```diff
@@ -2967,7 +2970,34 @@ int MotrStore::get_user_by_access_key(const DoutPrefixProvider *dpp, const std::
...
+  if (rc < 0){
+    ldout(cctx, 0) << "Failed to load user: rc = " << rc << dendl;
+    return rc;
+  }
+  u = new MotrUser(this, uinfo);
+  if (!u)
+    return -ENOMEM;
+
+  user->reset(u);  
+return 0;
```
[the remark this answers is not in the document]
Comment (shalakadharap) [reply src/rgw/rgw_sal_motr.cc:3000]: need help here. which tool to be used for this?

File: src/rgw/rgw_sal_motr.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:3032]: Because it will be returned at the next line anyway?

File: src/rgw/rgw_sal_motr.cc [hunk 11]
```diff
@@ -2967,6 +2970,33 @@ int MotrStore::get_user_by_access_key(const DoutPrefixProvider *dpp, const std::
...
 int MotrStore::get_user_by_email(const DoutPrefixProvider *dpp, const std::string& email, optional_yield y, std::unique_ptr<User>* user)
 {
+  int rc;
+  User *u;
+  bufferlist bl;
+  RGWUserInfo uinfo;
+  
+  rc = do_idx_op_by_name(RGW_IAM_MOTR_EMAIL_KEY,
+                           M0_IC_GET, email, bl);
+  if (rc < 0){
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:2980]: Style: add space between `){` here and in all other places, please.

**Review by andriytk — CHANGES_REQUESTED** [review]

**Comment by jjxsg** [comment]
Already addressed with https://github.com/Seagate/cortx-rgw/pull/101

**Pull request closed without merging**